src/software/ai/hl/stp/tactic/goalie/goalie_fsm.h

```diff
@@ -0,0 +1,448 @@
+        // then the ball must be either on a really sharp angle to the net where
+        // its impossible to get a shot, or the ball is behind the net, in which
+        // case we snap to either post
+        if (!clamped_goalie_pos)
+        {
+            if (ball.position().y() > 0)
+            {
+                goalie_pos =
+                    field.friendlyGoalpostPos() + Vector(0, -ROBOT_MAX_RADIUS_METERS);
+            }
+            else
+            {
+                goalie_pos =
+                    field.friendlyGoalpostNeg() + Vector(0, ROBOT_MAX_RADIUS_METERS);
+            }
+        }
+        else
+        {
+            goalie_pos = *clamped_goalie_pos;
+        }
+        return goalie_pos;
+    }
+
+    /**
+     * Gets intersections between the ball velocity ray and the full goal segment
+     *
+     * @param ball the ball to find interceptions with the full goal segment
+     * @param field the field to find interceptions on
+     *
+     * @return the intersections between the ball velocity ray and the full goal segment
+     */
+    static std::vector<Point> getIntersectionsBetweenBallVelocityAndFullGoalSegment(
+        const Ball &ball, const Field &field)
+    {
+        // compute intersection points from ball position and velocity
+        Ray ball_ray = Ray(ball.position(), ball.velocity());
+
+        // Create a segment along the goal line, slightly shortened to account for the
+        // robot radius so as we move along the segment we don't try to run into the goal
+        // posts. This will be used in case 3 as a fallback when we don't have an
+        // intersection with the crease lines
+        Segment full_goal_segment =
+            Segment(field.friendlyGoalpostNeg() + Vector(0, -ROBOT_MAX_RADIUS_METERS),
+                    field.friendlyGoalpostPos() + Vector(0, ROBOT_MAX_RADIUS_METERS));
+
+        return intersection(ball_ray, full_goal_segment);
+    }
+
+    /**
+     * Gets the "don't chip" rectangle inside the friendly defense area
+     *
+     * @param field the field to find the "don't chip" rectangle on
+     *
+     * @return the "don't chip" rectangle
+     */
+    static Rectangle getDontChipRectangle(const Field &field)
+    {
+        return Rectangle(
+            field.friendlyGoalpostNeg(),
+            field.friendlyGoalpostPos() + Vector(2 * ROBOT_MAX_RADIUS_METERS, 0));
+    }
+
+    /**
+     * Gets the point on the field from which the goalie should clear the ball
+     * @param field the field on which the goalie will clear the ball
+     * @param origin_opt the optional point to clear the ball from
+     *
+     * @return the position that the goalie should clear the ball from
+     */
+    static Point getClearOrigin(const Field &field, std::optional<Point> origin_opt)
+    {
+        Point clear_origin = field.friendlyGoalCenter() + Vector(0.5, 0);
+        if (origin_opt)
+        {
+            clear_origin = origin_opt.value();
```

Comment:
Mostly for testing purposes, I used this partly to test clearing from different locations/directions but mostly to justify/clarify why I validated a specific clear origin/direction (by overriding them specifically in the test case itself and using those values to validate the test, rather relying on the default value which isn't specified in the test case). I kept this as an optional parameter so that I only needed to set specific values when it was needed in a test

Reply:
So we wouldn't want to override the clear origin in actual gameplay? I'm a bit confused about why we'd want to test this functionality, then.

Reply:
It would still allow for freedom of choice in a game, but if we're okay with just the one default location every time then we wouldn't technically need this override option.. and also now since I'm probably going to change the clear origin to a location relative to the no-chip rectangle, I could make this location dependent on the goalie's location when it gets the ball and just have it go to the closest point to chip - so there wouldn't be anything that we need to manually set anymore, what do you think?

Reply:
that sounds good

Reply:
updated